Mathlib/NumberTheory/DirichletCharacter/Basic.lean

```diff
@@ -199,4 +209,21 @@ lemma conductor_one_dvd (n : ℕ) : conductor (1 : DirichletCharacter R 1) ∣ n
   rw [(isPrimitive_def _).mp isPrimitive_one_level_one]
   apply one_dvd _
 
+/-- The primitive character associated to a Dirichlet character. -/
+noncomputable def primitiveCharacter : DirichletCharacter R χ.conductor :=
+  Classical.choose (factorsThrough_conductor χ).choose_spec
+
+lemma reduction_isPrimitive : isPrimitive (χ.primitiveCharacter) := by
+  by_cases χ.conductor = 0
+  · rw [isPrimitive_def]
+    conv_rhs => rw [h]
+    rw [conductor_eq_zero_iff_level_eq_zero, h]
+  · refine' le_antisymm (Nat.le_of_dvd (Nat.pos_of_ne_zero h) (conductor_dvd_level _)) <|
+        conductor_le_conductor_mem_conductorSet <| conductor_mem_conductorSet χ
```

Comment:
```suggestion
  by_cases h : χ.conductor = 0
  · rw [isPrimitive_def]
    convert conductor_eq_zero_iff_level_eq_zero.mpr h
  · exact le_antisymm (Nat.le_of_dvd (Nat.pos_of_ne_zero h) (conductor_dvd_level _)) <|
      conductor_le_conductor_mem_conductorSet _ <| conductor_mem_conductorSet χ
```
(don't rely on the name `h` used by `by_cases` by default)

Reply:
I'm happy to change that, but why not rely on the default name?

Reply:
(The `_` before the last `<|` should go; in my test file I mistakenly had an explicit argument there.)
Relying on the deault name has the problem that a reader of the proof has to know about it (and the possible problem that the proof could break at some point if the default is changed).
In fact, I think that `by_cases` should use an inaccessible name by default, similarly to most other tactics.

Reply:
What you write about `by_cases` seems reasonable, I started a discussing about it here:

https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60by_cases.60using.20the.20default.20argument